kotlinlib/src/mill/kotlinlib/js/KotlinJSModule.scala

```diff
@@ -0,0 +1,509 @@
+      case OutputMode.KlibFile =>
+        Seq(
+          "-Xir-produce-klib-file",
+          "-ir-output-dir",
+          (destinationRoot / "libs").toIO.getAbsolutePath
+        )
+      case OutputMode.KlibDir =>
+        Seq(
+          "-Xir-produce-klib-dir",
+          "-ir-output-dir",
+          (destinationRoot / "classes").toIO.getAbsolutePath
+        )
+      case OutputMode.Js =>
+        Seq(
+          "-Xir-produce-js",
+          "-ir-output-dir",
+          (destinationRoot / "binaries").toIO.getAbsolutePath
+        )
+    }
+
+    innerCompilerArgs ++= outputArgs
+    innerCompilerArgs += s"-Xir-module-name=${moduleName()}"
+    innerCompilerArgs ++= (esTarget match {
+      case Some(x) => Seq("-target", x)
+      case None => Seq.empty
+    })
+
+    val compilerArgs: Seq[String] = Seq(
+      innerCompilerArgs.result(),
+      extraKotlinArgs,
+      // parameters
+      inputFiles
+    ).flatten
+
+    val compileDestination = os.Path(outputArgs.last)
+    if (irClasspath.isEmpty) {
+      T.log.info(
+        s"Compiling ${allKotlinSourceFiles.size} Kotlin sources to $compileDestination ..."
+      )
+    } else {
+      T.log.info(s"Linking IR to $compileDestination")
+    }
+    val workerResult = worker.compile(KotlinWorkerTarget.Js, compilerArgs: _*)
+
+    val analysisFile = T.dest / "kotlin.analysis.dummy"
+    if (!os.exists(analysisFile)) {
+      os.write(target = analysisFile, data = "", createFolders = true)
+    }
+
+    val artifactLocation = outputMode match {
+      case OutputMode.KlibFile => compileDestination / s"${moduleName()}.klib"
+      case OutputMode.KlibDir => compileDestination
+      case OutputMode.Js => compileDestination
+    }
+
+    workerResult match {
+      case Result.Success(_) =>
+        CompilationResult(analysisFile, PathRef(artifactLocation))
+      case Result.Failure(reason, _) =>
+        Result.Failure(reason, Some(CompilationResult(analysisFile, PathRef(artifactLocation))))
+      case e: Result.Exception => e
+      case Result.Aborted => Result.Aborted
+      case Result.Skipped => Result.Skipped
+    }
+  }
+
+  private def binaryKindToOutputMode(binaryKind: Option[BinaryKind]): OutputMode =
+    binaryKind match {
+      // still produce IR classes, but they won't be yet linked
+      case None => OutputMode.KlibDir
+      case Some(BinaryKind.Library) => OutputMode.KlibFile
+      case Some(BinaryKind.Executable) => OutputMode.Js
+    }
+
+  // these 2 exist to ignore values added to the display name in case of the cross-modules
+  // we already have cross-modules in the paths, so we don't need them here
+  private def moduleName() = millModuleSegments.value
+    .filter(_.isInstanceOf[Segment.Label])
+    .map(_.asInstanceOf[Segment.Label])
+    .last
+    .value
+
+  private def fullModuleName() = millModuleSegments.value
+    .filter(_.isInstanceOf[Segment.Label])
+    .map(_.asInstanceOf[Segment.Label].value)
+    .mkString("-")
```

Comment:
Where do these two names end up being used? If it's just for the compiled `.js` file, I wonder if we can get away with just providing a hardcoded name (IIRC ScalaJS just hardcodes it as `main.js`) and leaving the user to override and rename the file themselves if they need to

Reply:
I guess here comes the difference between the binary output of Scala.JS and Kotlin/JS: Kotlin/JS is able to split final `.js` files per module by using the [following flag](https://github.com/JetBrains/kotlin/blob/870720e57422ff0586dd251030fdd9dfaa445d4b/compiler/cli/cli-common/src/org/jetbrains/kotlin/cli/common/arguments/K2JSCompilerArguments.kt#L385), it can be seen [here](https://github.com/com-lihaoyi/mill/pull/3678/files#diff-237a14cc7b9dc8afcfb3c86aeb34224d77c2c0645771be9946cd019f529ce155R34-R49).

I guess the benefit of doing so lays mainly on the browser side, where such `.js` files can be loaded by the client in parallel.

Considering that, it is not possible to hardcode the default name (and I would avoid doing so), having the output matching the module name is more meaningful.